logserver.c

```diff
@@ -529,42 +777,53 @@ void pv_logserver_toggle(struct pantavisor *pv, const char *rev)
 		return;
 
 	if (pv_config_get_log_capture()) {
-		logserver_start(&logserver_g, rev);
+		logserver_start(rev);
 	} else
 		logserver_stop();
 }
 
 int pv_logserver_init()
 {
-	struct epoll_event ep_event;
-	char path[108]; // size of sockaddr_un sun_path
+	dl_list_init(&logserver_g.fdlst);
+	dl_list_init(&logserver_g.tmplst);
 
-	pv_paths_pv_file(path, PATH_MAX, LOGCTRL_FNAME);
-	logserver_g.sock_fd = logserver_open_socket(path);
-	logserver_g.epoll_fd = epoll_create1(0);
+	logserver_g.out.sfile = pv_config_get_log_server_output_single_file();
+	logserver_g.out.ftree = pv_config_get_log_server_output_file_tree();
+
+	errno = 0;
 
-	if (logserver_g.epoll_fd < 0 || logserver_g.sock_fd < 0) {
-		pv_log(DEBUG, "logserver_g epoll_fd = %d",
-		       logserver_g.epoll_fd);
-		pv_log(DEBUG, "logserver_g sock_fd = %d", logserver_g.sock_fd);
+	logserver_g.logsock = logserver_open_socket(LOGCTRL_FNAME, true);
+
+	if (logserver_g.out.sfile)
+		logserver_g.fdsock = logserver_open_socket(LOGFD_FNAME, true);
+
+	logserver_g.epfd = epoll_create1(0);
+
+	if (logserver_g.epfd < 0 || logserver_g.logsock < 0 ||
+	    (logserver_g.fdsock < 0 && logserver_g.out.sfile)) {
+		pv_log(DEBUG, "logserver_g epoll_fd = %d", logserver_g.epfd);
+		pv_log(DEBUG, "logserver_g log sock = %d", logserver_g.logsock);
+		pv_log(DEBUG, "logserver_g fd sock = %d", logserver_g.fdsock);
 		pv_log(DEBUG, "errno  =%d", errno);
 		goto out;
 	}
 
-	ep_event.events = EPOLLIN;
-	ep_event.data.fd = logserver_g.sock_fd;
-	if (epoll_ctl(logserver_g.epoll_fd, EPOLL_CTL_ADD, ep_event.data.fd,
-		      &ep_event))
+	if (logserver_epoll_add(logserver_g.logsock) == -1)
 		goto out;
 
-	logserver_start_service(&logserver_g, pv_bootloader_get_rev());
-	pv_log(DEBUG, "started log service with pid %d",
-	       (int)logserver_g.service_pid);
+	if (logserver_g.out.sfile) {
+		if (logserver_epoll_add(logserver_g.fdsock) == -1)
+			goto out;
+	}
+
+	logserver_start_service(pv_bootloader_get_rev());
+	pv_log(DEBUG, "started log service with pid %d", (int)logserver_g.pid);
 
 	return 0;
 out:
-	close(logserver_g.sock_fd);
-	close(logserver_g.epoll_fd);
+	close(logserver_g.logsock);
```

Comment:
maybe check here is these are > -1 to be closed

Reply:
Solved in 93c1fa8